algebra-core/src/curves/cuda/scalar_mul/cpu_gpu_macros.rs

```diff
@@ -0,0 +1,246 @@
+
+                    let (lower, upper) = res_ref.split_at_mut(num);
+                    res_ref = upper;
+                    let lower_exps = &exps_h_ref[..num];
+                    exps_h_ref = &exps_h_ref[num..];
+
+                    let mut table = DeviceMemory::<Self>::zeros(&ctx, num * Self::table_size());
+                    let mut exp = DeviceMemory::<u8>::zeros(&ctx, num * Self::num_u8());
+
+                    Self::generate_tables_and_recoding(lower, &mut table[..], lower_exps, &mut exp[..]);
+
+                    ctxs.push((device, ctx));
+                    bases_split.push(lower);
+                    tables.push(table);
+                    exps.push(exp);
+                };
+
+                println!("Split statically and allocated device: {}us", now.elapsed().as_micros());
+
+                rayon::scope(|s| {
+                    // Run jobs on GPUs
+                    for (i, (bases_gpu, time_gpu)) in bases_split.iter_mut().zip(times_gpu.iter_mut()).enumerate() {
+                        let n_gpu = n_gpus[i];
+                        let ctx = &ctxs[i].1;
+                        let table = &tables[i];
+                        let exp = &exps[i];
+
+                        s.spawn(move |_| {
+                            let now = std::time::Instant::now();
+                            let mut out = DeviceMemory::<Self>::zeros(ctx, n_gpu);
+                            P::scalar_mul_kernel(
+                                ctx,
+                                (n_gpu - 1) / cuda_group_size + 1, // grid
+                                cuda_group_size,     // block
+                                table.as_ptr(), exp.as_ptr(), out.as_mut_ptr(), n_gpu as isize
+                            )
+                            .expect("Kernel call failed");
+                            Self::batch_normalization(&mut out[..]);
+                            bases_gpu.clone_from_slice(&out.par_iter().map(|p| p.into_affine()).collect::<Vec<_>>()[..]);
+                            *time_gpu = now.elapsed().as_micros();
+                            println!("GPU {} finish", i);
+                        });
+                    }
+
+                    // Run on CPU
+                    s.spawn(|_| {
+                        let now = std::time::Instant::now();
+                        let exps_mut = &mut exps_h_ref.to_vec()[..];
+                        rayon::scope(|t| {
+                            for (b, s) in res_ref.chunks_mut(cpu_chunk_size).zip(exps_mut.chunks_mut(cpu_chunk_size)) {
+                                t.spawn(move |_| b[..].batch_scalar_mul_in_place(&mut s[..], 4));
+                            }
+                        });
+                        time_cpu = now.elapsed().as_micros();
+                        println!("CPU finish");
+                    });
+                });
+
+                // Update global microbenchmarking state
+                println!("old profile_data: {:?}", profile_data);
```

Comment:
This shouldn't appear in a normal run. Either change this to a debug print with logging or remove it.